Deprecated -t for --template to make it --tty

janetkuo opened the pull request:
Fixes #12248

janetkuo committed:
Deprecated -t for --template to make it --tty

JanetKuo commented:
cc @smarterclayton @bgrant0607 @thockin

bgrant0607 commented on docs/man/man1/kubectl-annotate.1:
```diff
@@ -63,8 +63,8 @@ resourcequotas (quota) or secrets.
 \fB\-t\fP, \fB\-\-template\fP=""
-    Template string or path to template file to use when \-o=template or \-o=templatefile.  The template format is golang templates [
-\[la]http://golang.org/pkg/text/template/#pkg-overview\[ra]]
+    Template string or path to template file to use when \-o=template or \-o=templatefile.  \-t is DEPRECATED to support \-\-tty.  The template format is golang templates [
```
I wouldn't mention that we plan to use -t for --tty -- that may just confuse users.

JanetKuo replied:
Fixed.

bgrant0607 commented:
cc @kubernetes/kubectl

smarterclayton commented:
No objections from me

deads2k commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -83,8 +86,11 @@ func OutputVersion(cmd *cobra.Command, defaultVersion string) string {
-	if len(outputFormat) == 0 && len(templateFile) != 0 {
-		outputFormat = "template"
+	if len(templateFile) != 0 {
+		glog.Warningf("%s is DEPRECATED and will be removed in a future version.", "-t")
```
Is there way for you to tell whether they've typed `-t` versus `--template`.  If I'm reading the issue right, the intent is to deprecate `-t`, but not `--template`.  I'm concerned about dumping this message too aggressively.

One alternative is to actually split `--template` into two flags and use `MarkDeprecated` for the `-t` flag.  If you do that, pflags will take care of the deprecation notification.

JanetKuo commented:
@deads2k Thanks for your suggestions. I didn't find a way that tells whether the users typed the shorthand or the flag name. Therefore, I updated the code to read the args and see if a deprecated shorthand, such as `-t`, is used, and then prints deprecated warning message. This way, we only output deprecated warning when `-t` is used (but not `--template`). (The code is a bit hacky though.) 

As for the alternative you mentioned, if we split the template flag, it'll be `--template` and `--t`, but not `-t`.

JanetKuo commented:
@k8s-bot test this please

deads2k commented on pkg/kubectl/cmd/util/factory.go:
```diff
@@ -303,6 +307,34 @@ func (f *Factory) BindFlags(flags *pflag.FlagSet) {
 	flags.SetNormalizeFunc(util.WordSepNormalizeFunc)
 }
 
+func (f *Factory) AddDeprecatedFlagShorthands(shorthand string) error {
```
How about moving this code into `pflag`?  It looks like this would be a good complementary method to https://github.com/kubernetes/kubernetes/blob/master/Godeps/_workspace/src/github.com/spf13/pflag/flag.go#L283 and detection you wrote below would function well inside https://github.com/kubernetes/kubernetes/blob/master/Godeps/_workspace/src/github.com/spf13/pflag/flag.go#L299 against the pre-normalized name.  The details (`flag.Deprecated`) could be used to explain that only `-t` is gone.

We have a couple committers on pflags.  @eparis Any objections to that approach instead of adding this here?

eparis replied:
I think that's a good idea to move this to pflag. Doing it "right" will take a bit more work. Obviously right now we only track if the short and long are both deprecated. So we'll need a way to say that only the short is deprecated. We really should start hiding the short from help text and documentation generation. I don't think `Set()` is actually the place you'll be able to find the short flag. I think it only handles long flags. You'll need to look at `origArg[0]` in `setFlag()`.

Also `CheckDeprecatedFlagShorthands()` isn't right...  It is legal (although silly) to do something like `-wt '{ name }'` (or something like there) where you have multiple shorthands after one `-`

JanetKuo replied:
Thanks @deads2k @eparis, I'll create a separate PR in github.com/spf13/pflag to support deprecating flag shorthands.

eparis commented:
@deads2k I also thought about the split idea but we'd need an options struct like we talked about yesterday....

ghodss commented:
Are we removing the golang template functionality altogether, or moving it to --golang-template or somehwere else? If we're moving it, it probably doesn't make sense to tell people it's being deprecated without giving them the future way of doing it.

JanetKuo commented:
@ghodss We're only deprecating `-t` but not `--template`.

ghodss commented:
@JanetKuo ah cool. I personally would recommend the help text be changed slightly to something like: `-t is DEPRECATED and will be removed in a future version - please use --template.` since to me the text reads more like the feature is deprecated.

JanetKuo commented on Godeps/_workspace/src/github.com/spf13/pflag/flag.go:
```diff
@@ -289,6 +290,16 @@ func (f *FlagSet) MarkDeprecated(name string, usageMessage string) error {
 }
 
+// Mark the shorthand of a flag deprecated in your program
+func (f *FlagSet) MarkShorthandDeprecated(name string, usageMessage string) error {
```
Code updated manually. It's easier to test the code change here. @deads2k @eparis please take a look and I'll apply this change to pflag package.

deads2k replied:
> Code updated manually. It's easier to test the code change here. @deads2k @eparis please take a look and I'll apply this change to pflag package.

The `pflags` change looks good.  Can you separate this into its own commit?  I'd suggest naming it something like "UPSTREAM: pflags 52: Add shorthand deprecator".  It makes it easier for people to find patches to bumped godeps.

JanetKuo commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -31,7 +31,11 @@ func AddPrinterFlags(cmd *cobra.Command) {
+	// TODO: remove template flag shorthand -t
+	cmd.Flags().StringP("template", "t", "", "Template string or path to template file to use when -o=template or -o=templatefile.  -t is DEPRECATED and will be removed in a future version, please use --template.  The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview].")
+	// TODO: uncomment this line after MarkShorthandDeprecated() is ready to be used
+	//	cmd.Flags().MarkShorthandDeprecated("template", "please use --template")
```
Here's how we'll deprecate shorthand in the updated pflag.

JanetKuo commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -31,7 +31,11 @@ func AddPrinterFlags(cmd *cobra.Command) {
 	cmd.Flags().StringP("output", "o", "", "Output format. One of: json|yaml|template|templatefile|wide.")
 	cmd.Flags().String("output-version", "", "Output the formatted object with the given version (default api-version).")
 	cmd.Flags().Bool("no-headers", false, "When using the default output, don't print headers.")
-	cmd.Flags().StringP("template", "t", "", "Template string or path to template file to use when -o=template or -o=templatefile.  The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview]")
```
@ghodss thanks. Help text updated as suggested.

JanetKuo commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -31,7 +31,11 @@ func AddPrinterFlags(cmd *cobra.Command) {
-	cmd.Flags().StringP("template", "t", "", "Template string or path to template file to use when -o=template or -o=templatefile.  The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview]")
+	// template shorthand -t is deprecated to support -t for --tty
+	// TODO: remove template flag shorthand -t
+	cmd.Flags().StringP("template", "t", "", "Template string or path to template file to use when -o=template or -o=templatefile.  -t is DEPRECATED and will be removed in a future version, please use --template.  The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview].")
```
@ghodss thanks. Help text updated as suggested.

eparis commented on Godeps/_workspace/src/github.com/spf13/pflag/flag.go:
```diff
@@ -566,9 +577,21 @@ func (f *FlagSet) setFlag(flag *Flag, value string, origArg string) error {
 	if len(flag.Deprecated) > 0 {
 		fmt.Fprintf(os.Stderr, "Flag --%s has been deprecated, %s\n", flag.Name, flag.Deprecated)
 	}
+	if len(flag.ShorthandDeprecated) > 0 && containsShorthand(origArg, flag.Shorthand) {
```
I'm concerned if I called `-abcd` and `c` was deprecated that I will get the error message 4 times...

JanetKuo replied:
I tested this with commands like `kubectl get pods -o template -wat {{.phase}}` and it only gives me one error message.

eparis commented on Godeps/_workspace/src/github.com/spf13/pflag/flag.go:
```diff
@@ -566,9 +577,21 @@ func (f *FlagSet) setFlag(flag *Flag, value string, origArg string) error {
+	arg = strings.SplitN(arg, "=", 2)[0]
+	return strings.Contains(arg, shorthand)
+}
+
```
Can you update `FlagUsages()` to hide the shorthand if it was deprecated?

JanetKuo replied:
Done

JanetKuo commented on Godeps/_workspace/src/github.com/spf13/pflag/flag.go:
```diff
@@ -386,12 +403,16 @@ func (f *FlagSet) FlagUsages() string {
+		if len(flag.ShorthandDeprecated) > 0 {
+			shorthand = ""
+		}
+		fmt.Fprintf(x, format, shorthand, flag.Name, flag.DefValue, flag.Usage)
```
Hide the shorthand if it's deprecated. (Since we're hiding it, we don't need to put the deprecation message in the help text.)

JanetKuo commented:
@k8s-bot test this please

deads2k commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -31,7 +31,11 @@ func AddPrinterFlags(cmd *cobra.Command) {
+	// TODO: remove template flag shorthand -t
+	cmd.Flags().StringP("template", "t", "", "Template string or path to template file to use when -o=template or -o=templatefile.  The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview].")
+	// TODO: uncomment this line after MarkShorthandDeprecated() is ready to be used
+	// cmd.Flags().MarkShorthandDeprecated("template", "please use --template")
```
I'm not certain of kube rules, but we'd accept this with the UPSTREAM patch/pull in openshift.  Add a better deprecation message and this lgtm.

JanetKuo replied:
@deads2k By setting usage message "`please use --template`", users will see "`Flag shorthand -t has been deprecated, please use --template`" when they use `-t`. Is it good enough?

janetkuo committed:
UPSTREAM: pflags 52: Add shorthand deprecator

eparis commented on docs/user-guide/kubectl/kubectl_annotate.md:
```diff
@@ -123,7 +123,7 @@ $ kubectl annotate pods foo description-
 * [kubectl](kubectl.md)	 - kubectl controls the Kubernetes cluster manager
 
-###### Auto generated by spf13/cobra at 2015-08-20 22:01:12.4790376 +0000 UTC
+###### Auto generated by spf13/cobra at 2015-08-20 23:24:35.129161073 +0000 UTC
```
Can you not make these changes?  (not really important)

JanetKuo replied:
This file is auto-generated by script so I cannot avoid this change.

eparis commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -31,7 +31,11 @@ func AddPrinterFlags(cmd *cobra.Command) {
+	// TODO: remove template flag shorthand -t
 	cmd.Flags().StringP("template", "t", "", "Template string or path to template file to use when -o=template, -o=templatefile or -o=jsonpath.  The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview]. The jsonpath template is composed of jsonpath expressions enclosed by {} [http://releases.k8s.io/HEAD/docs/user-guide/jsonpath.md]")
+	// TODO: uncomment this line after MarkShorthandDeprecated() is ready to be used
+	// cmd.Flags().MarkShorthandDeprecated("template", "please use --template instead")
```
And now you can use this  :)

JanetKuo replied:
Done

JanetKuo commented:
Shippable failed due to flake test https://github.com/kubernetes/kubernetes/pull/12202#issuecomment-133284488

deads2k commented:
Looks like you just need to update the generated docs and squash the non-upstream changes.

lgtm

deads2k commented:
lgtm.  @jlowdermilk can you tag?

JanetKuo commented:
@k8s-bot test this please

JanetKuo commented:
@k8s-bot test this please

JanetKuo commented:
@k8s-bot test this please

JanetKuo commented:
Restart flaky e2e test
@k8s-bot test this please

eparis commented:
Just to be clear. The LGTM label stands even after the rebase. Ready to merge.
